gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireContainer.java

```diff
@@ -491,9 +502,12 @@ public ExternalTableMetaData fetchHiveMetaData(boolean refresh) {
         schemaName = fullName.substring(0, schemaIndex);
         tableName = fullName.substring(schemaIndex + 1);
       }
+      ExternalCatalog extcat = Misc.getMemStore().getExternalCatalog();
+      if (extcat == null && Misc.initialDDLReplayInProgress()) {
```

Comment:
replay-in-progress flag would have been unset before external catalog initialization. Better to use a separate flag for indicating that catalog has been initialized or use a semaphore/notify-wait to wait rather than the loop. Also better to use an upper-limit on the wait (say 1 min) after which fail the operation with force reattempt just in case this node is stuck in external catalog initialization for some reason then entire cluster should not freeze.

Reply:
Oh yes. The ddlreplay is marked before hand itself. Good catch. For the hang concern I was planning to add checkCancelInProgress after every 20 sleeps. 1 minute of wait might be too less and made it infinite mainly because all the put operations are like indefinite wait, for whatever they are waiting, like primary selection etc, uness the node goes down.
But irrespective wait/notify can be done which will obviate all the above and add a finally block in postCreate to notify.